### PR TITLE
Migrate to use v4 version of module api

### PR DESCRIPTION
**Pull request by tsafin**
* fixed box_tuple_valdate
* bump branches version to v4
* use box_ibuf_t instead of struct ibuf
* and use luaT_toibuf instead of luaL_checkibuf

### CODE REVIEW

File: src/merger/merger.c [hunk 1]
```diff
@@ -494,6 +500,15 @@ luaL_merge_source_buffer_destroy(struct merge_source *base)
...
 
+static inline size_t
+ibuf_used(char *rpos, char *wpos)
+{
+	if (!rpos && !wpos)
+		return 0;
+	assert(wpos > rpos);
+	return wpos - rpos;
+}
+
```
Comment (Totktonada) [inline src/merger/merger.c:511]: There is 'Helpers' section above.

Comment (tsafin) [reply src/merger/merger.c:511]: Good point!

File: src/merger/merger.c [hunk 2]
```diff
@@ -494,6 +500,15 @@ luaL_merge_source_buffer_destroy(struct merge_source *base)
 	free(source);
 }
 
+static inline size_t
+ibuf_used(char *rpos, char *wpos)
+{
+	if (!rpos && !wpos)
+		return 0;
```
Comment (Totktonada) [inline src/merger/merger.c:507]: 1. Rule of thumb: explicitly check against NULL.
2. I would not just return 0 for an invalid buffer. Assertion fail is better than some subtle output data corruption.

**Review by Totktonada — COMMENTED** [review]
Math looks correct. LGTM.

File: src/merger/merger.c (hunk 2 above)
Comment (tsafin) [reply src/merger/merger.c]: 1. Done
2. Partly disagreed. `assert()` is only working for debug builds, but for release omitting sanity checks and blindly using incorrect data is worst approach for production. Collecting coredump is easy for debugging, but worse 100x times that check of input arguments and bailing out accordingly. 
In prior life (while being at Acronis and Ya.Taxi we always preferred both - `assert` and `if ... return`. Bceuase they are working at different times.

Comment (Totktonada) [reply src/merger/merger.c]: Let's set/return an error and handle it in all usages so.

**Pull request merged**